### PR TITLE
Added generator logic

### PR DESCRIPTION
**Pull request by josephlewis42**
Adds an MVP version of the generator: #1. With this, we get pretty nice support for most DevDocs items.

Known limitations:

- `third_party` assets are checked in and will need to be manually refreshed. Sadly, this info is only in the GitHub repo, but not the generated content on `devdocs.io` so there's a small window where it might become out of sync.
- Generation is very slow for resources like `ansible` which have very large number of entries in their navbars which causes Jinja to choke when instantiating and escaping tons of objects.
- Generating content with the default `libzim` compression is also slow: https://github.com/openzim/libzim/issues/544, we need some compression because images are base64 encoded and the navbars are repetitive, but doing so at the highest level is spending a bunch of time trying to further compress things with already high information density (e.g. JPEGs).

Once merged, good proof of concepts for this would be:

* https://github.com/openzim/zim-requests/issues/953
* https://github.com/openzim/zim-requests/issues/337

### CODE REVIEW

File: src/devdocs2zim/assets/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+These files are copied from DevDocs:
+
+    https://github.com/freeCodeCamp/devdocs
+
+devdocs_48.png is adapted from:
```
Comment (josephlewis42) [inline src/devdocs2zim/assets/README.md:5]: see if we can use common logic to resize rather than adapt

File: src/devdocs2zim/client.py [hunk 2]
```diff
@@ -94,6 +117,37 @@ class DevdocsIndexType(BaseModel):
     # Section slug. This appears to be unused.
     slug: str
 
+    def sort_precedence(self) -> SortPrecedence:
```
Comment (josephlewis42) [inline src/devdocs2zim/client.py:120]: test

File: src/devdocs2zim/client.py [hunk 3]
```diff
@@ -102,10 +156,36 @@ class DevdocsIndex(BaseModel):
     entries: list[DevdocsIndexEntry]
 
     # List of "types" or section headings.
-    # These are displayed mostly in order, except regular expressions are used to sort:
-    # https://github.com/freeCodeCamp/devdocs/blob/e28f81d3218bdbad7eac0540c58c11c7fe1d33d3/assets/javascripts/collections/types.js#L3
+    # These are displayed in the order they're found grouped by sort_precedence.
     types: list[DevdocsIndexType]
 
+    def build_navigation(self) -> list[NavigationSection]:
```
Comment (josephlewis42) [inline src/devdocs2zim/client.py:162]: test

File: src/devdocs2zim/client.py [hunk 4]
```diff
@@ -74,14 +87,24 @@ class DevdocsIndexEntry(BaseModel):
     path: str
 
     # Name of the type (section) the entry is located under.
-    type: str
+    type: str | None
```
Comment (josephlewis42) [inline src/devdocs2zim/client.py:90]: note about nones

File: src/devdocs2zim/client.py [hunk 5]
```diff
@@ -102,10 +156,36 @@ class DevdocsIndex(BaseModel):
...
+        """Builds a navigation hierarchy that's soreted correctly for rendering."""
+
+        sections_by_precedence: dict[SortPrecedence, list[DevdocsIndexType]] = (
+            defaultdict(list)
+        )
+        for section in self.types:
+            sections_by_precedence[section.sort_precedence()].append(section)
+
+        links_by_section_name: dict[str, list[DevdocsIndexEntry]] = defaultdict(list)
+        for entry in self.entries:
```
Comment (josephlewis42) [inline src/devdocs2zim/client.py:172]: itertools.groupby?

File: src/devdocs2zim/client.py [hunk 6]
```diff
@@ -102,10 +156,36 @@ class DevdocsIndex(BaseModel):
...
+    # These are displayed in the order they're found grouped by sort_precedence.
     types: list[DevdocsIndexType]
 
+    def build_navigation(self) -> list[NavigationSection]:
+        """Builds a navigation hierarchy that's soreted correctly for rendering."""
+
+        sections_by_precedence: dict[SortPrecedence, list[DevdocsIndexType]] = (
+            defaultdict(list)
+        )
+        for section in self.types:
```
Comment (josephlewis42) [inline src/devdocs2zim/client.py:168]: itertools.groupby

File: src/devdocs2zim/generator.py [hunk 7]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+            loader=FileSystemLoader(ROOT_DIR.joinpath("templates")),
+            autoescape=select_autoescape(),
+        )
+
+        self.page_template = self.env.get_template("page.html")  # type: ignore
+        self.license_template = self.env.get_template("license.html")  # type: ignore
+
+        self.logo_path = self.asset_path("devdocs_48.png")
+        self.copyright_path = self.asset_path("COPYRIGHT")
+        self.license_path = self.asset_path("LICENSE")
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:305]: assert initializes

File: src/devdocs2zim/generator.py [hunk 8]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+        """Returns the path to name in the assets folder.
+
+        Raises ValueError if the resource doesn't exist.
+        """
+        path = ROOT_DIR.joinpath("assets", name)
+        if not path.exists():
+            raise ValueError(f"File not found at {path}")
+        return path
+
+    def load_common_files(self) -> list[StaticItem]:
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:318]: test

File: src/devdocs2zim/generator.py [hunk 9]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+
         # List all docs and copy one by one
         for doc_metadata in selected_doc_metadata:
-            logger.info(f"Fetching {doc_metadata.slug}")
+            self.generate_zim(
+                doc_metadata,
+                common_resources,
+            )
+
+    def generate_zim(
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:370]: test e2e

File: src/devdocs2zim/generator.py [hunk 10]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+        # navigation item that opens the page.
+        page_to_title: dict[str, str] = {}
+        for entry in index.entries:
+            path_without_fragment = entry.path_without_fragment
+            if path_without_fragment == entry.path:
+                page_to_title[path_without_fragment] = entry.name
+            elif path_without_fragment not in page_to_title:
+                page_to_title[path_without_fragment] = entry.name
+
+        # Explicitly inject the index.
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:455]: note why we have to do this

File: src/devdocs2zim/generator.py [hunk 11]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+        page_to_title["index"] = f"{doc_metadata.name} Documentation"
+
+        nav_sections = index.build_navigation()
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:458]: rename navigation

File: src/devdocs2zim/generator.py [hunk 12]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+        # Check formatting early to bail if any templates are invalid.
+        for doc_metadata in selected_doc_metadata:
+            self.zim_config.format(doc_metadata.placeholders())
+
+        common_resources = self.load_common_files()
+
         # List all docs and copy one by one
         for doc_metadata in selected_doc_metadata:
-            logger.info(f"Fetching {doc_metadata.slug}")
+            self.generate_zim(
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:365]: Add TODO for progress

File: src/devdocs2zim/generator.py [hunk 13]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
 
-        raise NotImplementedError("ZIM creation is not yet implemented")
+        logger.info("  Adding common resources...")
+        for resource in common_resources:
+            creator.add_item(resource)  # type: ignore
+
+        # Set the title for each page to the navigation item that opens the page
+        # to the top i.e. without a fragment if it exists. Otherwise, the first
+        # navigation item that opens the page.
+        page_to_title: dict[str, str] = {}
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:447]: extract and test

File: src/devdocs2zim/generator.py [hunk 14]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+                page_to_title[path_without_fragment] = entry.name
+
+        # Explicitly inject the index.
+        page_to_title["index"] = f"{doc_metadata.name} Documentation"
+
+        nav_sections = index.build_navigation()
+
+        logger.info(f"  Rendering {len(page_to_title)} pages...")
+        counter = 0
+        render_delta = 0.0
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:462]: remove timers

File: tests/test_basic.py [hunk 15]
```diff
@@ -0,0 +1,7 @@
+# pyright: strict, reportUnusedExpression=false
```
Comment (josephlewis42) [inline tests/test_basic.py:1]: drop

File: src/devdocs2zim/templates/page.html [hunk 16]
```diff
@@ -0,0 +1,47 @@
+{# Devdocs is an SPA so each page will have (nearly) identical content. #}
```
Comment (josephlewis42) [inline src/devdocs2zim/templates/page.html:1]: add note about all the `| safes`

File: src/devdocs2zim/templates/license.html [hunk 17]
```diff
@@ -0,0 +1,17 @@
+{# Devdocs is an SPA so each page will have (nearly) identical content. #}
```
Comment (josephlewis42) [inline src/devdocs2zim/templates/license.html:1]: Fix header content

File: src/devdocs2zim/templates/page.html (hunk 16 above)
Comment (josephlewis42) [inline src/devdocs2zim/templates/page.html:1]: Add note about CSS

File: src/devdocs2zim/generator.py [hunk 18]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+        creator = Creator(zim_path, "index")
+        creator.config_metadata(
+            Name=formatted_config.name_format,
+            Title=formatted_config.title_format,
+            Publisher=formatted_config.publisher,
+            Date=datetime.datetime.now(tz=datetime.UTC).date(),
+            Creator=formatted_config.creator,
+            Description=formatted_config.description_format,
+            LongDescription=formatted_config.long_description_format,
+            Language=LANGUAGE_ISO_639_3,
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:394]: Why constant?

File: src/devdocs2zim/generator.py [hunk 19]
```diff
@@ -274,15 +291,215 @@ def __init__(
...
+    def generate_zim(
+        self, doc_metadata: DevdocsMetadata, common_resources: list[StaticItem]
+    ):
+        """Generates a zim for a single document."""
+        logger.info(f"Generating ZIM for {doc_metadata.slug}")
+
+        formatted_config = self.zim_config.format(doc_metadata.placeholders())
+        zim_path = Path(self.output_folder, f"{formatted_config.name_format}.zim")
+
+        if zim_path.exists():
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:379]: Note about clobber

File: src/devdocs2zim/client.py (hunk 6 above)
Comment (josephlewis42) [reply src/devdocs2zim/client.py:168]: doesn't make this cleaner, ignoring

File: src/devdocs2zim/generator.py [hunk 20]
```diff
@@ -274,15 +289,216 @@ def __init__(
...
+            self.add_zim_contents(
+                creator=started_creator,
+                doc_metadata=doc_metadata,
+                index=index,
+                db=db,
+                common_resources=common_resources,
+            )
+
+    @staticmethod
+    def page_titles(pages: list[DevdocsIndexEntry]) -> dict[str, str]:
```
Comment (josephlewis42) [inline src/devdocs2zim/generator.py:422]: test

File: src/devdocs2zim/assets/COPYRIGHT [hunk 21]
```diff
@@ -0,0 +1,13 @@
+Copyright 2013-2024 Thibaut Courouble and other contributors
```
Comment (josephlewis42) [inline src/devdocs2zim/assets/COPYRIGHT:1]: move to third_party/devdocs

**Comment by benoit74** [comment]
Hi @josephlewis42, thank you very much, I will review this tomorrow, I missed the notification in the avalanche I got when coming back from holidays, sorry for that!

**Review by benoit74 — CHANGES_REQUESTED** [review]
See few inline comments

I've created a ZIM of lua~5.4 and cpp. I will open separate issues to discuss what I've found there (nothing which is a blocker to merge this code, more stuff to be discussed / fixed before releasing a 0.1.0 version)

Regarding the time it takes to complete, I'm not that surprised. cpp tooks 15 minutes to complete, which is OK. We are used to scraper taking hours or even sometime days to complete, so this is not a concern from my PoV. But still good to mention and try to enhance, the faster the scraper can be the better.

Regarding known limitations, I suggest we open one issue per limitation so that we can discuss separately how to address them (or even decide to not address them ^^). I agree we do not need to fix them before the MVP.

File: src/devdocs2zim/generator.py [hunk 22]
```diff
@@ -274,15 +289,227 @@ def __init__(
...
+                # Compression is needed because images are embedded as Base64 and
+                # navigation is similar across pages.
+                should_compress=True,
+                mimetype="text/html",
+            )
+
+            # Tracking metadta
+            counter += 1
+            if counter % 100 == 0:
+                logger.debug(f"  Progress {counter} / {len(page_to_title)} pages ")
```
[the remark this answers is not in the document]
Comment (josephlewis42) [reply src/devdocs2zim/generator.py:513]: That works for me! We can also emit the JSON progress data for the overall progress but I thought that would be a little too much to do in this PR.

File: src/devdocs2zim/generator.py [hunk 23]
```diff
@@ -274,15 +289,227 @@ def __init__(
...
+                content=db.get(path, MISSING_PAGE),
+            )
+            creator.add_item_for(  # type: ignore
+                path,
+                title=title,
+                content=page_content,  # type: ignore
+                is_front=True,
+                # Compression is needed because images are embedded as Base64 and
+                # navigation is similar across pages.
+                should_compress=True,
```
[the remark this answers is not in the document]
Comment (josephlewis42) [reply src/devdocs2zim/generator.py:506]: I think I made a bad assumption here, the Godot tutorial which had lots of images is slow, but those images aren't in-lined (tracking that in #21). I had opened the file in Kiwix not realizing that Kiwix will load web resources so I assumed everything was in that `db.json` file that DevDocs caches in LocalStorage.

Instead, I think the biggest contributor is the navbar. As far as I can tell, the entries need to be rendered as relative links on every page (until we use JS) which can easily double the size of every page.

An alternative could be to just render the section the user is in and link section titles to the first page in that section.

File: src/devdocs2zim/generator.py [hunk 24]
```diff
@@ -1,20 +1,35 @@
...
+    DevdocsIndexEntry,
     DevdocsMetadata,
 )
-from devdocs2zim.constants import logger
+from devdocs2zim.constants import LANGUAGE_ISO_639_3, NAME, ROOT_DIR, VERSION, logger
+
+# Content to display for pages missing from DevDocs.
+MISSING_PAGE = (
+    "<h2>This documentation is missing.</h2>"
+    "<p>This is an error with DevDocs, not your ZIM reader e.g. Kiwix.</p>"
```
[the remark this answers is not in the document]
Comment (josephlewis42) [reply src/devdocs2zim/generator.py:31]: SGTM, I also added a warning when this happens. I only saw it happen once in the Bash documentation and I think it's fixed now.

File: src/devdocs2zim/generator.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (benoit74) [reply src/devdocs2zim/generator.py:517]: Let's live with it for now, and will see how this moves forward in next iterations

**Review by benoit74 — APPROVED** [review]

**Pull request merged**